frame/election-provider-multi-phase/test-staking-e2e/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg(test)]
+mod mock;
+
+pub(crate) const LOG_TARGET: &str = "tests::epm";
+
+use mock::*;
+use sp_npos_elections::{to_supports, StakedAssignment};
+use sp_runtime::Perbill;
+
+use crate::mock::RuntimeOrigin;
+
+// syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		log::$level!(
+			target: crate::LOG_TARGET,
+			concat!("🛠️  ", $patter)  $(, $values)*
+        )
+    };
+}
+
+fn log_current_time() {
+	log!(
+		trace,
+		"block: {:?}, session: {:?}, era: {:?}, EPM phase: {:?} ts: {:?}",
+		System::block_number(),
+		Session::current_index(),
+		Staking::current_era(),
+		ElectionProviderMultiPhase::current_phase(),
+		Timestamp::now()
+	);
+}
+
+#[test]
+fn setup_works() {
+	ExtBuilder::default().initialize_first_session(true).build_and_execute(|| {
+		assert_eq!(active_era(), 0);
+
+		assert!(start_next_active_era().is_ok());
+		assert!(start_next_active_era().is_ok());
+		assert_eq!(active_era(), 2);
+
+		// if the solution is delayed, EPM will end up in emergency mode and eras won't progress.
+		assert!(start_next_active_era_delayed_solution().is_ok());
```

Comment:
I am confused. If solution is delayed, it should not go to next era right?

Reply:
Oh `start_active_era` returns okay if `delay_solution` is set. Was a little unintuitive though. I was expecting it would throw error since it could not start new era.